### PR TITLE
System.Net.Sockets - Tests - Removed more blocking on APM tests

### PR DESCRIPTION
**Pull request by Drawaes**

### CODE REVIEW

**Comment by Drawaes** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop OSX x64 Debug Build please

**Comment by stephentoub** [comment]
@Drawaes, I've not seen this error before:
```
2017-12-02 23:39:45,871: INFO: proc(54): run_and_log_output: Output:    System.Net.Http.Functional.Tests.ManagedHandler_HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test.SetDelegate_ConnectionSucceeds(acceptedProtocol: Tls, requestOnlyThisProtocol: False) [FAIL]
2017-12-02 23:39:45,873: INFO: proc(54): run_and_log_output: Output:       System.Net.Http.HttpRequestException : An error occurred while sending the request.
2017-12-02 23:39:45,873: INFO: proc(54): run_and_log_output: Output:       ---- System.IO.IOException : The decryption operation failed, see inner exception.
2017-12-02 23:39:45,874: INFO: proc(54): run_and_log_output: Output:       -------- Interop+OpenSsl+SslException : Decrypt failed with OpenSSL error - SSL_ERROR_SSL.
2017-12-02 23:39:45,874: INFO: proc(54): run_and_log_output: Output:       ------------ System.Security.Cryptography.CryptographicException : Error occurred during a cryptographic operation.
2017-12-02 23:39:45,882: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2017-12-02 23:39:46,122: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs(437,0): at System.Net.Http.HttpConnection.<SendAsync>d__38.MoveNext()
2017-12-02 23:39:46,122: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,122: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionHandler.cs(43,0): at System.Net.Http.HttpConnectionHandler.<SendAsyncWithAwaitedConnection>d__4.MoveNext()
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Managed/CookieHandler.cs(31,0): at System.Net.Http.CookieHandler.<SendAsync>d__3.MoveNext()
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs(62,0): at System.Net.Http.AuthenticateAndRedirectHandler.<SendAsync>d__7.MoveNext()
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/HttpClient.cs(464,0): at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext()
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.AcceptAllCerts.cs(54,0): at System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test.<>c__DisplayClass3_1.<<SetDelegate_ConnectionSucceeds>b__0>d.MoveNext()
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,123: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/Common/tests/System/Net/Http/LoopbackServer.cs(68,0): at System.Net.Test.Common.LoopbackServer.<>c__DisplayClass3_0.<CreateServerAsync>b__0(Task t)
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          /root/coreclr/src/mscorlib/shared/System/Threading/ExecutionContext.cs(151,0): at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          /root/coreclr/src/mscorlib/src/System/Threading/Tasks/Task.cs(2440,0): at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.AcceptAllCerts.cs(52,0): at System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test.<SetDelegate_ConnectionSucceeds>d__3.MoveNext()
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          ----- Inner Stack Trace -----
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs(285,0): at System.Net.Security.SslStreamInternal.<ReadAsyncInternal>d__32`1.MoveNext()
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,124: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs(957,0): at System.Net.Http.HttpConnection.<FillAsync>d__64.MoveNext()
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs(866,0): at System.Net.Http.HttpConnection.<ReadNextLineAsync>d__62.MoveNext()
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          --- End of stack trace from previous location where exception was thrown ---
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs(312,0): at System.Net.Http.HttpConnection.<SendAsync>d__38.MoveNext()
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          ----- Inner Stack Trace -----
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs(272,0): at Interop.OpenSsl.Decrypt(SafeSslHandle context, Byte[] outBuffer, Int32 offset, Int32 count, SslErrorCode& errorCode)
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs(207,0): at System.Net.Security.SslStreamPal.EncryptDecryptHelper(SafeDeleteContext securityContext, ReadOnlyMemory`1 input, Int32 offset, Int32 size, Boolean encrypt, Byte[]& output, Int32& resultSize)
2017-12-02 23:39:46,125: INFO: proc(54): run_and_log_output: Output:          ----- Inner Stack Trace -----
```
It's obviously not related to your test changes here, but could it be related to any of your recent changes related to OpenSSL, SslStream, etc.?

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs [hunk 1]
```diff
@@ -354,63 +354,43 @@ public void Socket_BeginConnectV4IPAddressToV4Host_Throws()
...
+        private async Task DualModeBeginConnect_IPAddressToHost_Helper(IPAddress connectTo, IPAddress listenOn, bool dualModeServer)
         {
-            int port;
             using (Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp))
-            using (SocketServer server = new SocketServer(_log, listenOn, dualModeServer, out port))
+            using (SocketServer server = new SocketServer(_log, listenOn, dualModeServer, out int port))
             {
-                IAsyncResult async = socket.BeginConnect(connectTo, port, null, null);
-                socket.EndConnect(async);
+                await Task.Factory.FromAsync(socket.BeginConnect(connectTo, port, null, null), socket.EndConnect);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: FWIW, this overload of FromAsync is much less efficient than if you used the one that took two delegates:
```C#
await Task.Factory.FromAsync(socket.BeginConnect, socket.EndConnect, port, null);
```
as the overload you've used needs to use ThreadPool.RegisterWaitForSingleObject.  Not a big deal for test code.

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs [hunk 2]
```diff
@@ -479,44 +445,40 @@ public class DualModeBeginConnect : DualModeBase
         [Theory]
         [MemberData(nameof(DualMode_IPAddresses_ListenOn_DualMode_Data))]
         [PlatformSpecific(TestPlatforms.Windows)]  // Connecting sockets to DNS endpoints via the instance Connect and ConnectAsync methods not supported on Unix
-        private void DualModeBeginConnect_IPAddressListToHost_Helper(IPAddress[] connectTo, IPAddress listenOn, bool dualModeServer)
+        private async Task DualModeBeginConnect_IPAddressListToHost_Helper(IPAddress[] connectTo, IPAddress listenOn, bool dualModeServer)
         {
             int port;
             using (Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp))
             using (SocketServer server = new SocketServer(_log, listenOn, dualModeServer, out port))
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: Nit: you changed earlier occurrences of `out port` to `out int port`, but not this one.

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs [hunk 3]
```diff
@@ -479,44 +445,40 @@ public class DualModeBeginConnect : DualModeBase
...
 
         [Theory]
         [MemberData(nameof(DualMode_Connect_IPAddress_DualMode_Data))]
         [PlatformSpecific(TestPlatforms.Windows)]  // Connecting sockets to DNS endpoints via the instance Connect and ConnectAsync methods not supported on Unix
-        public void DualModeBeginConnect_LoopbackDnsToHost_Helper(IPAddress listenOn, bool dualModeServer)
+        public async Task DualModeBeginConnect_LoopbackDnsToHost_Helper(IPAddress listenOn, bool dualModeServer)
         {
             int port;
             using (Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp))
             using (SocketServer server = new SocketServer(_log, listenOn, dualModeServer, out port))
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: Ditto.

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs [hunk 4]
```diff
@@ -436,37 +416,23 @@ public void Socket_BeginConnectV4IPEndPointToV4Host_Throws()
...
-        private void DualModeBeginConnect_IPEndPointToHost_Helper(IPAddress connectTo, IPAddress listenOn, bool dualModeServer)
+        private async Task DualModeBeginConnect_IPEndPointToHost_Helper(IPAddress connectTo, IPAddress listenOn, bool dualModeServer)
         {
-            int port;
             using (Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp))
-            using (SocketServer server = new SocketServer(_log, listenOn, dualModeServer, out port))
+            using (SocketServer server = new SocketServer(_log, listenOn, dualModeServer, out int port))
             {
-                IAsyncResult async = socket.BeginConnect(new IPEndPoint(connectTo, port), null, null);
-                socket.EndConnect(async);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: In removing this blocking, do we still have coverage of said blocking?  We shouldn't remove all of it: we still need to test the path where EndXxx is called in this pattern, blocking waiting for BeginConnect to complete when no callback is provided.

**Comment by Drawaes** [comment]
The error above, it's possible that not clearing the errors constantly has knock on issue in other crypto functions. What this might be showing, is that errors are getting left in the queues and now there is no hard "clear" it's showing up a problem where previously by luck it was getting the queues cleared. I need to take a look at the function that blew.

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs (hunk 4 above)
Comment (Drawaes) [reply src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: That is true, we don't get hard blocking tests by removing them. On the flipside previously there was no callback testing :) I could add a single blocking test to cover just the blocking functionality and leave the rest as call backs?

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs (hunk 3 above)
Comment (Drawaes) [reply src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: I removed anywhere in the file there was a warning for inline warnings, it was easier to make the file self consistent.

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs (hunk 1 above)
Comment (Drawaes) [reply src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: I have changed all of these. Might as well make the tests run as fast as possible seeing as there are like 12m on a outerloop.

**Comment by Drawaes** [comment]
I can't see any error details here

https://mc.dot.net/#/user/Drawaes/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/9a38c63dcbe93190d58bc25c3c559f2a5c0e01a8/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.ManagedHandler_HttpClientHandler_DangerousAcceptAllCertificatesV...~2FSetDelegate_ConnectionSucceeds

While its not related to this I want to investigate

**Comment by Drawaes** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please

**Comment by stephentoub** [comment]
@Drawaes, I don't know why that page is blank (cc: @dotnet/dnceng), but you can see the wire in the log at https://mc.dot.net/#/user/Drawaes/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/9a38c63dcbe93190d58bc25c3c559f2a5c0e01a8/workItem/System.Net.Http.Functional.Tests/wilogs

File: src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs (hunk 4 above)
Comment (stephentoub) [reply src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs]: Ok

**Comment by stephentoub** [comment]
@drawaes, here's another similar failure on a different PR:
https://mc.dot.net/#/user/dotnet-maestro-bot/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/ec0e83a3a9bb2b05e289c3aec374aa879e8691a0/workItem/System.Net.Http.Functional.Tests/wilogs

If we don't figure this out soon, we should roll back that change, assuming it's the likely cause.

**Comment by Drawaes** [comment]
Roll it back if you like, I can add it to my testing PR and look at it there if that is easier

**Comment by MattGal** [comment]
The reason the page is blank is that we truncate both type and method names that exceed 100 chars, and System.Net.Http.Functional.Tests.ManagedHandler_HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test is 110 chars long.  It's a performance limitation of storing our test results in SQL.

This truncation should only happen on theory parameters in the near future;  however I've filed https://github.com/dotnet/core-eng/issues/2209 to make sure we look at this in the new non-SQL result storage.

**Comment by stephentoub** [comment]
@dotnet-bot test Linux x64 Release Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please

**Pull request merged**